File: sdk-py/src/openglass/crypto/hash.py
"""``H(bytes) = sha256(bytes)`` (SPEC §7.1)."""

from __future__ import annotations

import hashlib
import re

_HEX_RE = re.compile(r"^[0-9a-f]+$")


def sha256(data: bytes) -> bytes:
    return hashlib.sha256(data).digest()


def to_hex(data: bytes) -> str:
    return data.hex()


def hex_to_bytes(value: str) -> bytes:
    if not _HEX_RE.match(value) or len(value) % 2 != 0:
        raise ValueError(f"hex_to_bytes: not a valid lowercase hex string: {value!r}")
    return bytes.fromhex(value)

File: sdk-js/src/crypto/hash.ts
import { createHash } from "node:crypto";

/** H(bytes) = sha256(bytes) (SPEC §7.1). */
export function sha256(bytes: Uint8Array): Buffer {
  return createHash("sha256").update(bytes).digest();
}

export function hex(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("hex");
}

export function hexToBytes(value: string): Buffer {
  if (!/^[0-9a-f]+$/.test(value) || value.length % 2 !== 0) {
    throw new Error(`hexToBytes: not a valid lowercase hex string: ${value}`);
  }
  return Buffer.from(value, "hex");
}
